src/libcore/str/mod.rs

```diff
@@ -34,99 +36,33 @@ use result::Result::{self, Ok, Err};
 use slice::{self, SliceExt};
 use usize;
 
-pub use self::pattern::Pattern;
-pub use self::pattern::{Searcher, ReverseSearcher, DoubleEndedSearcher, SearchStep};
+pub mod pattern;
 
-mod pattern;
+macro_rules! iter_methods {
+    (: $te:ty) => {
```

Comment:
Hm yes I do agree that not having these implementations drift apart is a good idea, I'll give this another pass with an eye to see if I have a suggestion or two for the macros themselves.